File: XRayne.Dashboard/src/routes/nodes/ui/create-node-dialog/lib/node-provision-step.ts
import type { NodeProvisionState } from "@features/node";

const NODE_PROVISION_STEPS = [
  "queued",
  "preparing",
  "uploading",
  "installing",
  "installingDependencies",
  "downloadingImage",
  "configuringCertificate",
  "startingContainer",
  "verifying",
  "completed",
  "failed",
] as const;

export type NormalizedNodeProvisionStep = (typeof NODE_PROVISION_STEPS)[number];

export function getNodeProvisionStep(
  step: NodeProvisionState["step"],
): NormalizedNodeProvisionStep {
  if (typeof step === "number") {
    return NODE_PROVISION_STEPS[step] ?? "queued";
  }

  return step;
}

File: Infrastructure/States/Enums/NodeProvisionStep.cs
using System.Text.Json.Serialization;

namespace Infrastructure.States;

[JsonConverter(typeof(JsonStringEnumConverter))]
public enum NodeProvisionStep
{
    [JsonStringEnumMemberName("queued")]
    Queued,

    [JsonStringEnumMemberName("preparing")]
    Preparing,

    [JsonStringEnumMemberName("uploading")]
    Uploading,

    [JsonStringEnumMemberName("installing")]
    Installing,

    [JsonStringEnumMemberName("installing_dependencies")]
    InstallingDependencies,

    [JsonStringEnumMemberName("downloading_image")]
    DownloadingImage,

    [JsonStringEnumMemberName("starting_container")]
    StartingContainer,

    [JsonStringEnumMemberName("installing_core")]
    InstallingCore,

    [JsonStringEnumMemberName("verifying")]
    Verifying,

    [JsonStringEnumMemberName("completed")]
    Completed,

    [JsonStringEnumMemberName("failed")]
    Failed
}
